server/src/main/java/org/elasticsearch/common/geo/TriangleTreeWriter.java

```diff
@@ -0,0 +1,618 @@
+                        ab = ca;
+                        cX = bX;
+                        cY = bY;
+                        ca = bc;
+                        bX = tempX;
+                        bY = tempY;
+                        bc = tempBool;
+                    }
+                }
+            }
+            this.aX = aX;
+            this.aY = aY;
+            this.bX = bX;
+            this.bY = bY;
+            this.cX = cX;
+            this.cY = cY;
+            this.ab = ab;
+            this.bc = bc;
+            this.ca = ca;
+            this.minX = aX;
+            this.maxX = Math.max(aX, Math.max(bX, cX));
+            this.minY = Math.min(aY, Math.min(bY, cY));
+            this.maxY = Math.max(aY, Math.max(bY, cY));
+            type = TYPE.TRIANGLE;
+        }
+
+        private static List<TriangleTreeLeaf> fromPoints(CoordinateEncoder encoder, Point... points) {
+            List<TriangleTreeLeaf> triangles = new ArrayList<>(points.length);
+            for (int i = 0; i < points.length; i++) {
+                triangles.add(new TriangleTreeLeaf(encoder.encodeX(points[i].getX()), encoder.encodeY(points[i].getY())));
+            }
+            return triangles;
+        }
+
+        private static List<TriangleTreeLeaf> fromRectangle(CoordinateEncoder encoder, Rectangle... rectangles) {
+            List<TriangleTreeLeaf> triangles = new ArrayList<>(2 * rectangles.length);
+            for (Rectangle r : rectangles) {
+                triangles.add(new TriangleTreeLeaf(
+                    encoder.encodeX(r.getMinX()), encoder.encodeY(r.getMinY()), true,
+                    encoder.encodeX(r.getMaxX()), encoder.encodeY(r.getMinY()), false,
+                    encoder.encodeX(r.getMinX()), encoder.encodeY(r.getMaxY()), true));
+                triangles.add(new TriangleTreeLeaf(
+                    encoder.encodeX(r.getMinX()), encoder.encodeY(r.getMaxY()), false,
+                    encoder.encodeX(r.getMaxX()), encoder.encodeY(r.getMinY()), true,
+                    encoder.encodeX(r.getMaxX()), encoder.encodeY(r.getMaxY()), true));
+            }
+            return triangles;
+        }
+
+        private static List<TriangleTreeLeaf> fromLine(CoordinateEncoder encoder, Line line) {
+            List<TriangleTreeLeaf> triangles = new ArrayList<>(line.length() - 1);
+            for (int i = 0, j = 1; i < line.length() - 1; i++, j++) {
+                triangles.add(new TriangleTreeLeaf(encoder.encodeX(line.getX(i)), encoder.encodeY(line.getY(i)),
+                    encoder.encodeX(line.getX(j)), encoder.encodeY(line.getY(j))));
+            }
+            return triangles;
+        }
+
+        private static List<TriangleTreeLeaf> fromPolygon(CoordinateEncoder encoder, Polygon polygon) {
+            // TODO: We are going to be tessellating the polygon twice, can we do something?
+            // TODO: Tessellator seems to have some reference to the encoding but does not need to have.
```

Comment:
it may be possible to re-use the tessellations that were built by the LuceneGeometryIndexer by passing it in AbstractGeometryFieldMapper#parse because the shape is indexed/tessellated before the doc values.

Reply:
Not at the moment. The tessellation happens inside Lucene so we do not have access to that. Maybe is something I can push forward there.